packages/astro/src/content/vite-plugin-content-imports.ts

```diff
@@ -41,116 +45,201 @@ export function astroContentImportPlugin({
+								contentEntryExts.some((ext) => modUrl.endsWith(ext))
+							) {
+								const mod = await viteServer.moduleGraph.getModuleByUrl(modUrl);
+								if (mod) {
+									viteServer.moduleGraph.invalidateModule(mod);
+								}
 							}
 						}
 					}
+				});
+			},
+			async transform(code, id) {
+				if (isContentFlagImport(id, contentEntryExts)) {
+					// Escape before Rollup internal transform.
+					// Base on MUCH trial-and-error, inspired by MDX integration 2-step transform.
+					return { code: escapeViteEnvReferences(code) };
 				}
-			});
+			},
 		},
-		async transform(code, id) {
-			if (isContentFlagImport(id, contentEntryExts)) {
-				// Escape before Rollup internal transform.
-				// Base on MUCH trial-and-error, inspired by MDX integration 2-step transform.
-				return { code: escapeViteEnvReferences(code) };
+	];
+
+	if (settings.contentEntryTypes.some((t) => t.getRenderModule)) {
+		plugins.push({
+			name: 'astro:content-render-imports',
+			async load(viteId) {
+				if (!contentEntryExts.some((ext) => viteId.endsWith(ext))) return;
+
+				const { fileId } = getFileInfo(viteId, settings.config);
+				for (const contentEntryType of settings.contentEntryTypes) {
+					if (contentEntryType.getRenderModule) {
+						const entry = await getContentEntryModuleFromCache(fileId);
+						// Cached entry must exist (or be in-flight) when importing the module via content collections.
+						// This is ensured by the `astro:content-imports` plugin.
+						if (!entry)
+							throw new AstroError({
+								...AstroErrorData.UnknownContentCollectionError,
+								message: `Unable to render ${JSON.stringify(
+									fileId
+								)}. Did you import this module directly without using a content collection query?`,
+							});
+
+						return contentEntryType.getRenderModule({ entry });
+					}
+				}
+			},
+		});
+	}
+
+	// Used by the `render-module` plugin to avoid double-parsing your schema
+	const contentEntryModuleByIdCache = new Map<string, ContentEntryModule | 'loading'>();
+	const awaitingCacheById = new Map<string, ((val: ContentEntryModule) => void)[]>();
+	function getContentEntryModuleFromCache(id: string) {
+		const value = contentEntryModuleByIdCache.get(id);
+		// It's possible for Vite to load modules that depend on this cache
+		// before the cache is populated. In that case, we queue a promise
+		// to be resolved by `setContentEntryModuleCache`.
+		if (value === 'loading') {
+			return new Promise<ContentEntryModule>((resolve, reject) => {
+				const awaiting = awaitingCacheById.get(id) ?? [];
+				awaiting.push(resolve);
+				awaitingCacheById.set(id, awaiting);
+			});
+		} else if (value) {
+			return Promise.resolve(value);
+		}
+		return Promise.resolve(undefined);
+	}
+
+	async function setContentEntryModuleCache({
+		fileId,
+		pluginContext,
+	}: {
+		fileId: string;
+		pluginContext: PluginContext;
+	}): Promise<ContentEntryModule> {
+		contentEntryModuleByIdCache.set(fileId, 'loading');
```

Comment:
I admit that having a "loading" value inside a cache is not something you see often. Usually, you either have or do not have value. Would you mind explaining how this caching works?

Reply:
Yep, sorry for the unconventional implementation! Trying to track 3 possible states here:
1. Module has not been cached yet (`undefined`)
2. Module is being resolved, but the async operation hasn't completed yet (`'loading'`)
3. Module has been resolved and is stored (`ContentEntryModule`)

This gets around an annoying problem I hit in production builds, where Vite tries to load a content entry's rendered content _before_ its frontmatter module has been resolved. By tracking which cached entries haven't resolved yet with `'loading'`, we can queue up a promise to "wait" for that frontmatter module to be resolved before rendering the post's content module. I tried implementing this with cached promises instead, but truthfully couldn't wrap my head around the complexity for a friday evening 😅 

If this is too confusing, I'm happy to refactor. Let me know if there's an implementation I'm not seeing